docs/proposals/202003_thanos_rules_federation.md

```diff
@@ -0,0 +1,226 @@
+
+1. the rule type (recording rule vs. alerting rule)
+2. the rule name
+3. the rule label names
+4. the rule expression `expr` field
+5. the alerting rule `for` field
+
+are being used as the deduplication identifier. Disjunct entries are simply merged by adding them to the result set.
+
+Thanos Querier presents the result of the fan-out on a `/api/v1/rules` endpoint which is compatible with the Prometheus' rules API endpoint.
+Additionally Thanos Querier gains a new `--rule.replica-label` command line argument which falls back to `--query.replica-label` if unset. The replica labels refer to the same labels as specified in the `external_labels` section of Prometheus and the `--label` command line argument of Thanos Ruler.
+
+#### Examples
+
+A stream of alerting rules is merged from different Thanos Ruler API instances.
+These may be remote Thanos Ruler, Prometheus Sidecar, or even Thanos Querier instances.
+The merging Thanos Querier has `--rule.replica-label=replica`
+
+Scenario 1:
+
+As specified, the rule type and then rule name is used for deduplication.
+
+Given the following stream of incoming rule groups and containing recording/alerting rules:
+```text
+group: a
+   recording:<name:"a1" last_evaluation:<seconds:1 > >
+   alert:    <name:"r1" last_evaluation:<seconds:1 > >
+group: b
+   recording:<name:"r1" last_evaluation:<seconds:1 > >
+group: a
+   recording:<name:"r2" last_evaluation:<seconds:1 > >
+```
+
+The output becomes:
+```text
+group: a
+   alert:    <name:"a1" last_evaluation:<seconds:1 > >
+   recording:<name:"r1" last_evaluation:<seconds:1 > >
+   recording:<name:"r2" last_evaluation:<seconds:1 > >
+group: b
+   recording:<name:"r1" last_evaluation:<seconds:1 > >
+```
+
+Note in the example above how the recording rule `r1` is not deduplicated as it is contained in two different groups.
+
+Scenario 2:
+
+The next level of deduplication is governed by the label/value set of the underlying recording/alerting rule while respecting the replica label.
+For a given conflict, the youngest is preferred. For alerting rules, the youngest firing rule is preferred.
+
+Given the following stream of incoming recording rules:
+```text
+group: a
+   recording:<name:"r1" labels:<labels:<name:"replica" value:"thanos-ruler-1" > > last_evaluation:<2006-01-02T10:00:00> >
+group: a
+   recording:<name:"r1" labels:<labels:<name:"replica" value:"thanos-ruler-2" > > last_evaluation:<2006-01-02T10:01:00> >
+```
+
+The output becomes:
+```text
+group: a
+   recording:<name:"r1" labels:<labels:<name:"replica" value:"thanos-ruler-2" > > last_evaluation:<2006-01-02T10:01:00> >
+```
+
+Given the following stream of incoming alerting rules:
+```text
+group: a
+   alert:<state:FIRING name:"a1" labels:<labels:<name:"replica" value:"thanos-ruler-1" > > last_evaluation:<2006-01-02T10:00:00> >
+group: a
+   alert:<             name:"a1" labels:<labels:<name:"replica" value:"thanos-ruler-2" > > last_evaluation:<2006-01-02T10:01:00> >
```

Comment:
Where has the state disappeared? :smile: 

```suggestion
   alert:<state:PENDING name:"a1" labels:<labels:<name:"replica" value:"thanos-ruler-2" > > last_evaluation:<2006-01-02T10:01:00> >
```

Reply:
sounds good, added, thank you!